File: 2019/14/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n");

function convertString(str) {
  const [n, chem] = str.split(" ");
  return [chem, Number(n)];
}

class DefaultDict extends Map {
  constructor(defaultValueFunc, entries) {
    super(entries);
    this.defaultValueFunc = defaultValueFunc;
  }

  get(key) {
    if (!this.has(key)) {
      const defaultValue = this.defaultValueFunc(key);
      this.set(key, defaultValue);
      return defaultValue;
    }
    return super.get(key);
  }
}

function Recipe(quantity, inputs) {
  this.quantity = quantity;
  this.inputs = inputs.map(convertString);
}

function parseInput(input) {
  const recipes = new Map();

  for (const line of input) {
    const [inputs, output] = line.split(" => ");
    const [chemical, quantity] = convertString(output);
    recipes.set(chemical, new Recipe(quantity, inputs.split(", ")));
  }

  return recipes;
}

function calculateOreReq(chemical, quantity, recipes, surplus) {
  if (chemical === "ORE") {
    return quantity;
  }

  const available = surplus.get(chemical);
  const used = Math.min(surplus.get(chemical), quantity);
  surplus.set(chemical, available - used);
  quantity -= used;
  if (quantity === 0) {
    return 0;
  }

  const recipe = recipes.get(chemical);
  const batches = Math.ceil(quantity / recipe.quantity);
  surplus.set(
    chemical,
    surplus.get(chemical) + batches * recipe.quantity - quantity
  );
  let ore = 0;

  for (const [inputChemical, inputQuantity] of recipe.inputs) {
    ore += calculateOreReq(
      inputChemical,
      inputQuantity * batches,
      recipes,
      surplus
    );
  }

  return ore;
}

function part1(recipes) {
  const surplus = new DefaultDict(() => 0);
  return calculateOreReq("FUEL", 1, recipes, surplus);
}

function part2(recipes, oreAvailable) {
  let lo = 0;
  let hi = oreAvailable;

  while (lo < hi) {
    const mid = Math.floor((lo + hi + 1) / 2);
    const surplus = new DefaultDict(() => 0);
    const ore = calculateOreReq("FUEL", mid, recipes, surplus);
    if (ore > oreAvailable) {
      hi = mid - 1;
    } else {
      lo = mid;
    }
  }

  return lo;
}

const recipes = parseInput(input);
console.log(part1(recipes));
console.log(part2(recipes, 1000000000000));

File: 2019/14/solution.py
from collections import defaultdict
import math
import os

def parse_input(input_lines):
    reactions = {}
    for line in input_lines:
        inputs, output = line.split(" => ")
        output_qty, output_chem = output.split(" ")
        inputs = [(int(qty), chem) for qty, chem in (component.split(" ") for component in inputs.split(", "))]
        reactions[output_chem] = (int(output_qty), inputs)
    return reactions

def ore_required(reactions, target_chemical, target_amount, surplus):
    if target_chemical == 'ORE':
        return target_amount

    if target_chemical in surplus:
        available = min(surplus[target_chemical], target_amount)
        surplus[target_chemical] -= available
        target_amount -= available

    if target_amount == 0:
        return 0

    output_qty, inputs = reactions[target_chemical]
    batches = math.ceil(target_amount / output_qty)
    surplus[target_chemical] += batches * output_qty - target_amount

    ore_needed = 0
    for input_qty, input_chem in inputs:
        ore_needed += ore_required(reactions, input_chem, batches * input_qty, surplus)

    return ore_needed

def part1(reactions):
    surplus = defaultdict(int)
    return ore_required(reactions, 'FUEL', 1, surplus)

def ore_required_for_fuel(reactions, fuel_amount, surplus):
    return ore_required(reactions, 'FUEL', fuel_amount, surplus)

def part2(reactions, total_ore):
    low = 0
    high = total_ore

    while low < high:
        mid = (low + high + 1) // 2
        surplus = defaultdict(int)
        required_ore = ore_required_for_fuel(reactions, mid, surplus)

        if required_ore <= total_ore:
            low = mid
        else:
            high = mid - 1

    return low

if __name__ == "__main__":
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    input_lines = [line.strip() for line in file]
  reactions = parse_input(input_lines)
  print(part1(reactions))
  print(part2(reactions, 1000000000000))
